src/useFormo.ts

```diff
@@ -522,144 +515,187 @@ export function useFormo<
     V extends ArrayRecord<Values, SK>
   >(
     values: Values
-  ): Record<K, TaskEither<NonEmptyArray<FieldError>, any>> =>
-    pipe(
-      arrayValues(values),
-      record.mapWithIndex((name, subforms) =>
-        pipe(
-          subforms,
-          array.traverseWithIndex(
-            taskEither.getTaskValidation(
-              nonEmptyArray.getSemigroup<FieldError>()
-            )
-          )((index) => validateSubform<SK, K, V>(values as V, index, name))
+  ): Promise<Record<K, Result<NonEmptyArray<FieldError>, any>>> {
+    const arrValues = arrayValues(values);
+    const results: Record<
+      K,
+      Result<NonEmptyArray<FieldError>, any>
+    > = {} as any;
+    for (const name_ in arrValues) {
+      const name: K = name_ as K;
+      const subforms = arrValues[name];
+      const arrayResult = await Promise.all(
+        subforms.map((_, index) =>
+          validateSubform<SK, K, V>(values as V, index, name)
         )
-      )
-    );
+      );
+      const arrayFailures = arrayResult
+        .filter(isFailure)
+        .flatMap((r) => r.failure);
+      if (arrayFailures.length > 0) {
+        results[name] = failure(arrayFailures as NonEmptyArray<FieldError>);
+      } else {
+        results[name] = success(
+          arrayResult.filter(isSuccess).map((r) => r.success)
+        );
+      }
+    }
+    return results;
+  }
 
-  const validateAllFields = (
+  async function validateAllFields(
     values: Values
-  ): TaskEither<
-    NonEmptyArray<FieldError>,
-    ValidatedValues<Values, Validators, ArrayValidators>
-  > =>
-    pipe(
-      values,
-      record.mapWithIndex((name) => validateField(name, values)),
-      (fieldValidations) =>
-        sequenceS(
-          taskEither.getTaskValidation(nonEmptyArray.getSemigroup<FieldError>())
-        )({
-          ...fieldValidations,
-          ...validateAllSubforms(values),
-        } as any)
-    ) as any;
-
-  const setAllTouched = () => {
-    setTouched(
-      (pipe(values, record.map(constTrue)) as unknown) as Partial<Touched>
-    );
-    pipe(
-      arrayValues(values),
-      record.mapWithIndex((field, v) =>
-        pipe(
-          v,
-          array.mapWithIndex((index, r) =>
-            pipe(
-              r,
-              record.mapWithIndex((subfield) =>
-                dispatch({
-                  type: "setFieldArrayTouched",
-                  field,
-                  index,
-                  subfield,
-                  touched: true,
-                })
-              )
-            )
-          )
-        )
-      )
-    );
+  ): Promise<
+    Result<
+      NonEmptyArray<FieldError>,
+      ValidatedValues<Values, Validators, ArrayValidators>
+    >
+  > {
+    let failures: Array<FieldError> = [];
+
+    const plainValues = {} as Partial<
+      ValidatedValues<Values, Validators, ArrayValidators>
+    >;
+    for (const name in values) {
+      const result = await validateField(name, values);
```

Comment:
could/should be a `Promise.all` instead of loop, so as to run multiple async validations in parallel

Reply:
same as above